File: Typescript/Node/src/algos.ts

let n= process.argv.length > 2 ? +process.argv[2] : 1000000

let startTS = Date.now()
let primes = findPrimes(n)
let stopTS = Date.now()
let duration = stopTS - startTS
let primesCount = primes.length
console.log(`Typescript: Found ${primes.length} primes until ${n} in ${duration}ms`)

let startRecTS = Date.now()
let primesRec = findPrimesRec(n);
let stopRecTS = Date.now()
let durationRec = stopRecTS - startRecTS
let primesRecCount = primesRec.length
console.log(`Typescript (rec): Found ${primesRec.length} primes until ${n} in ${durationRec}ms`)


function findPrimes(n: number): Array<number> {
    let a = new Array<boolean>(n);
    a.fill(false)
    let primes = [2]
    for (let i = 3; i < n; i += 2) {
        if (!a[i]) {
            primes.push(i);
            for (let j = i * i; j < n; j += 2 * i) {
                a[j] = true;
            }
        }
    }
    return primes;
}

// recursive implementation to test tail recusion optimization
function findPrimesRec(n: number): Array<number> {
    function filterSieve(p: number, a: Array<boolean>): Array<boolean> {
        for (let i = p * p; i < n; i += 2 * p) {
            a[i] = false
        }
        return a
    }

    function sieve(p: number, a: Array<boolean>): Array<number> {
        switch (true) {
            case (p * p > n):
                {
                    let primes = [2]
                    for (let i = 3; i < n; i += 2) {
                        if (a[i]) {
                            primes.push(i)
                        }
                    }
                    return primes
                }
            case (a[p]):
                return sieve((p + 2), filterSieve(p, a))
            default:
                return sieve((p + 2), a)
        }
    }

    let a = new Array<boolean>(n);
    a.fill(true)

    for (let i = 4; i < n; i += 2) {
        a[i] = false
    }
    return sieve(3, a)
}

File: Rust/mre_algos/src/algos.rs
pub fn find_primes(n: usize) -> Vec<usize> {
    let mut a = vec![false; n];
    let mut primes = Vec::new();
    primes.push(2);
    for i in (3..n).step_by(2) {
        if !a[i] {
            primes.push(i);
            for i in (i * i..n).step_by(2 * i) {
                a[i] = true;
            }
        }
    }
    primes
}

// recursive implementation to test tail recusion optimization
pub fn find_primes_rec(n: usize) -> Vec<usize> {
    fn sieve(n: usize, p: usize, a: &mut Vec<bool>) -> Vec<usize> {
        if p * p > n {
            a.iter()
                .skip(2)
                .enumerate()
                .filter(|(_i, x)| **x)
                .map(|(i, _x)| i)
                .collect()
        } else if a[p] {
            for i in (p * p..n).step_by(2 * p) {
                a[i] = false;
            }
            sieve(n, p + 2, a)
        } else {
            sieve(n, p + 2, a)
        }
    }
    let mut a = vec![true; n];
    for i in (4..n - 1).step_by(2) {
        a[i] = false;
    }
    sieve(n - 1, 3, &mut a)
}
